media-proxy/src/session-base.cc

```diff
@@ -0,0 +1,211 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <string.h>
+#include <strings.h>
+#include <sys/stat.h>
+#include <arpa/inet.h>
+
+#include <mcm_dp.h>
+
+#include "session-base.h"
+#include <unistd.h>
+
+static int on_connect_callback_wrapper(memif_conn_handle_t conn, void *priv_data)
+{
+    session *connection = (session *)priv_data;
+    return connection->on_connect_cb(conn);
+}
+
+static int on_disconnect_callback_wrapper(memif_conn_handle_t conn, void *priv_data)
+{
+    session *connection = (session *)priv_data;
+    return connection->on_disconnect_cb(conn);
+}
+
+static int on_receive_callback_wrapper(memif_conn_handle_t conn, void *priv_data, uint16_t qid)
+{
+    session *connection = (session *)priv_data;
+    return connection->on_receive_cb(conn, qid);
+}
+
+void session::memif_event_loop()
+{
+    int err;
+    memif_socket_handle_t memif_socket = memif_conn_args.socket;
+
+    do {
```

Comment:
I think we should implement graceful exit here. Currently we use pthread_cancel(), which could potentially create some problems as this thread handles events and calls functions like on_receive_cb(). Those functions lock mutexes, and interact with libfabric and MTL. A lot of funny staff can happen when this is interrupted in random places.

We could set a little shorter timeout then infinity and check some new `stop` flag on every loop.

Reply:
Good idea!

Reply:
I suggest to do it with std::jthread, std::stop_source and std::stop_token

Reply:
https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/245

Reply:
memif_stop added in https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/240/commits/e1a5864a0a6f12017139869a11f5089261321e18